### PR TITLE
Feat: find text via images

### LINKED ISSUE #99: feat: image to text

### PR DESCRIPTION
**Pull request by joschkabraun**
closes #99

### CODE REVIEW

**Review by makram93 — APPROVED** [review]
LGTM! Maybe, as you suggested, adding an integration test for lyrics as well would be helpful

File: now/frontend/frontend.py [hunk 1]
```diff
@@ -183,13 +186,30 @@ def convert_file_to_document(query):
             for m in matches
             if m.scores['cosine'].value > st.session_state.min_confidence
         ]
+        # if output_modality == 'text':
+        #     longest_text = max(map(len, [match.text for match in matches]))
```
Comment (makram93) [inline now/frontend/frontend.py:190]: Also, can you remove the commented code if it is not needed, else you can leave a small comment

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -135,7 +135,7 @@ jobs:
           pytest --suppress-no-test-exit-code --cov=now --cov-report=xml \
             -v -s -m "not gpu" ${{ matrix.test-path }}
           echo "::set-output name=codecov_flag::now"
-        timeout-minutes: 45
+        timeout-minutes: 90
```
[the remark this answers is not in the document]
Comment (joschkabraun) [reply .github/workflows/ci.yml:138]: Actually no, I will revert. Thank you!

File: now/data_loading/build_datasets.py [hunk 3]
```diff
@@ -473,6 +475,166 @@ def transform(d: Document):
...
+        lambda x: len(x['Lyric'].split('\n')), axis=1
+    )
+    lyrics = lyrics.query('num_sentences <= 100')
+
+    lyrics = pd.merge(lyrics, artists_df, left_on='ALink', right_on='Link')
+
+    lyrics = lyrics[lyrics['Genres'].str.contains(genre)]
+
+    if 0 < max_size:
+        lyrics = lyrics.sample(frac=1)
```
[the remark this answers is not in the document]
Comment (joschkabraun) [reply now/data_loading/build_datasets.py:508]: It shuffles the dataset such that we randomly pick max_size sentences.

File: now/frontend/frontend.py [hunk 4]
```diff
@@ -124,7 +124,10 @@ def convert_file_to_document(query):
         '<style>div.st-bf{flex-direction:column;} div.st-ag{font-weight:bold;padding-right:50px;}</style>',
         unsafe_allow_html=True,
     )
-    media_type = st.radio('', ["text", "Image"], on_change=clear_match)
+    if output_modality == 'image':
+        media_type = st.radio('', ["Text", "Image"], on_change=clear_match)
+    elif output_modality == 'text':
+        media_type = st.radio('', ["Image", "Text"], on_change=clear_match)
```
[the remark this answers is not in the document]
Comment (joschkabraun) [reply now/frontend/frontend.py:130]: For searching in text, I think searching with images should be the default option as searching text-to-text will be less used.

[the remark this answers is not in the document]
Comment (makram93) [reply now/frontend/frontend.py:130]: okay, that makes sense!

**Comment by florian-hoenicke** [comment]
@joschkabraun Is it possible to use custom text by now or will this be handled in a follow-up ticket?

**Comment by florian-hoenicke** [comment]
When trying to run it locally I get the following error:
```
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│                                                                                                  │
│ /Users/florianhonicke/jina/now/now/cli/__init__.py:125 in <module>                               │
│                                                                                                  │
│   122                                                                                            │
│   123                                                                                            │
│   124 if __name__ == '__main__':                                                                 │
│ ❱ 125 │   cli()                                                                                  │
│   126                                                                                            │
│ /Users/florianhonicke/jina/now/now/cli/__init__.py:121 in cli                                    │
│                                                                                                  │
│   118 │   args['kind_path'] = kind_path                                                          │
│   119 │   from now.run_all_k8s import run_k8s                                                    │
│   120 │                                                                                          │
│ ❱ 121 │   run_k8s(os_type=os_type, arch=arch, **args)                                            │
│   122                                                                                            │
│   123                                                                                            │
│   124 if __name__ == '__main__':                                                                 │
│                                                                                                  │
│ /Users/florianhonicke/jina/now/now/run_all_k8s.py:62 in run_k8s                                  │
│                                                                                                  │
│   59 │   │   │   │   gateway_port,                                                               │
│   60 │   │   │   │   gateway_host_internal,                                                      │
│   61 │   │   │   │   gateway_port_internal,                                                      │
│ ❱ 62 │   │   │   ) = run_backend.run(user_input, is_debug, tmpdir, **kwargs)                     │
│   63 │   │   │   frontend_host, frontend_port = run_frontend.run(                                │
│   64 │   │   │   │   output_modality=user_input.output_modality,                                 │
│   65 │   │   │   │   dataset=user_input.dataset,                                                 │
│                                                                                                  │
│ /Users/florianhonicke/jina/now/now/run_backend.py:151 in run                                     │
│                                                                                                  │
│   148 │   │   embedding_size=embedding_size,                                                     │
│   149 │   │   tmpdir=tmpdir,                                                                     │
│   150 │   │   finetuning=finetuning,                                                             │
│ ❱ 151 │   │   **kwargs,                                                                          │
│   152 │   )                                                                                      │
│   153 │   return gateway_host, gateway_port, gateway_host_internal, gateway_port_internal        │
│   154                                                                                            │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
TypeError: deploy_flow() missing 1 required positional argument: 'output_modality'
```

I selected:
- 📝 text
- 🎤 rap lyrics (200K docs)
- 🐣 create new
- 📍 local (Kubernetes in Docker)

**Comment by florian-hoenicke** [comment]
Well done! @joschkabraun

**Pull request merged**